Fix multiple copies of components on discard changing a move operation.

mushtaqak opened the pull request:
This PR fixes multiple copies of components on discard changing a move operation.
[Sandbox](https://studio-movecomponent.sandbox.edx.org/home/)

This is the after PR of move feature. There is an issue when we move a component from A to B location and then discard the changes in A, component would be present in A and B both locations. This PR helps in preventing from that situation.

# [TNL-6670](https://openedx.atlassian.net/browse/TNL-6670) [TNL-6726](https://openedx.atlassian.net/browse/TNL-6726)

mushtaqak commented:
jenkins run a11y

mushtaqak commented:
jenkins run bokchoy

mushtaqak commented:
jenkins run js

mushtaqak commented:
jenkins run js

muzaffaryousaf commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -733,8 +733,8 @@ def _move_item(source_usage_key, target_parent_usage_key, user, target_index=Non
-        elif source_parent.location == target_parent.location:
-            error = _('You can not move an item into the same parent.')
+        elif source_parent.location == target_parent.location or source_item.location in target_parent.children:
+            error = _('Item is already present in target location.')
```
What about moving into same parent. Don't we need to show `You can not move an item into the same parent.` in that case ?

mushtaqak replied:
It's actually same message, I just had to make this so that both cases could be handled.

muzaffaryousaf commented on common/lib/xmodule/xmodule/modulestore/mongo/draft.py:
```diff
@@ -793,19 +793,26 @@ def delete_draft_only(root_location):
+                version
+                for version in versions_found
+                if version.get('_id').get('revision') != MongoRevisionKey.draft
+            ][0]
```
Why 0th only ?

mushtaqak replied:
previously 0th versions_found was being used, comments explained that 0th would be publish and 1st would be draft, and on the next line it was being asserted that the versions_found[0] is not the draft in case it was fetched, so I just refactored that logic

doctoryes replied:
In the old version of the code, an empty `versions_found` would exit the `delete_draft_only` method - this change would cause an error instead, which seems undesirable.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/tests/test_item.py:
```diff
@@ -1144,6 +1154,58 @@ def test_move_logging(self, mock_logger):
+    @ddt.data(ModuleStoreEnum.Type.mongo, ModuleStoreEnum.Type.split)
+    def test_move_and_discard_changes(self, store_type):
+        """
+        Test that discard changes operation works as expected after a move operation bring back the component back.
```
`Verifies that discard changes operation brings moved component back to source location and remove the component from target location`.

mushtaqak replied:
Perfect docstring, I will replace with this one.

muzaffaryousaf commented on common/lib/xmodule/xmodule/modulestore/mongo/draft.py:
```diff
@@ -793,19 +793,26 @@ def delete_draft_only(root_location):
+                    # delete_subtree as soon as we find an item with a draft version, if there is only 1 version
+                    # it must be published (since adding a child to a published item creates a draft of the parent).
+                    # Also, when an item is moved, it also creates a draft just like addition and deletion.
+                    delete_draft_only(Location.from_deprecated_string(child_loc))
```
can we replace this with current method ?

mushtaqak commented:
@nasthagiri Can you take a look at this ?

mushtaqak commented:
@doctoryes Can you take a quick look at this ?

doctoryes commented on common/lib/xmodule/xmodule/modulestore/draft_and_published.py:
```diff
@@ -134,6 +134,38 @@ def _flag_publish_event(self, course_key):
+            return
+        item_parent_location = unicode(item.parent.for_branch(None))
+        if item_parent_location and item_parent_location != unicode(source_parent_location):
+            # If an item parent is different than the current parent then it means it is moved.
```
Change to:
` Item's parent is different than its new parent - so it has moved.`

doctoryes commented on common/lib/xmodule/xmodule/modulestore/draft_and_published.py:
```diff
@@ -134,6 +134,38 @@ def _flag_publish_event(self, course_key):
+        item_parent_location = unicode(item.parent.for_branch(None))
+        if item_parent_location and item_parent_location != unicode(source_parent_location):
+            # If an item parent is different than the current parent then it means it is moved.
+            # Remove item from the list of children of moved parent children.
```
`Remove item from the current parent's children list.`

doctoryes commented on common/lib/xmodule/xmodule/modulestore/draft_and_published.py:
```diff
@@ -134,6 +134,38 @@ def _flag_publish_event(self, course_key):
+            user_id (int)                                   : User id
+
+        Returns:
+           location                                         : Locator of item.
```
Or `None` if the item was not moved.

doctoryes commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split_draft.py:
```diff
@@ -446,6 +446,10 @@ def copy_from_published(root_block_id):
+                # If block is not the main container being discarded.
+                if root_block_id != BlockKey.from_usage_key(location):
+                    block_key = location.course_key.make_usage_key(root_block_id.type, root_block_id.id)
+                    self.remove_reference_if_moved(block_key, location, user_id)
```
Is this code intentionally placed after the recursive call so that all the `_update_block_in_structure` calls are made before the `remove_reference_if_moved` calls? Or not? Are there any implications to this ordering?

doctoryes commented:
@bmedx FYI.

mushtaqak commented:
@muzaffaryousaf @doctoryes @bmedx This is ready for review again. I have updated the sandbox.

P.S:  I have added the latest commit that simplifies the logic and add it at a common place.

mushtaqak commented:
@doctoryes This PR fixes an unexpected scenario in discard changes operation after a move operation has been attempted - more on  https://openedx.atlassian.net/browse/TNL-6670

TLDR;
> There is an issue when we move a component from A to B location and then discard the changes in A, component would be present in A and B both locations. This PR helps in preventing from such situation.

mushtaqak commented:
jenkins run python

mushtaqak commented:
@doctoryes @bmedx Can you review this today ?

doctoryes commented on common/lib/xmodule/xmodule/modulestore/draft_and_published.py:
```diff
@@ -134,37 +138,58 @@ def _flag_publish_event(self, course_key):
+
+        # Add item to new parent at particular location.
+        if source_item.location not in new_parent_item.children:
+            insert_at = insert_at if insert_at is not None else len(new_parent_item.children)
```
Is `insert()` below zero-based? If so, this code is off-by-one. Couldn't you just do a `new_parent_item.children.append()` if insert_at is None?

mushtaqak replied:
append would always add the item  in the last, so we are inserting an item at a particular location.

doctoryes replied:
Right - that's true if `insert_at` is not None. If it *is* None, the code attempts to insert the item at the end of the children list. `append` would be simpler for that purpose.

doctoryes commented on common/lib/xmodule/xmodule/modulestore/draft_and_published.py:
```diff
@@ -134,37 +138,58 @@ def _flag_publish_event(self, course_key):
+            log.info(
+                '%s added to %s children',
+                unicode(source_item.location),
+                unicode(old_parent_item.location)
```
`old_parent_item` => `new_parent_item`

doctoryes commented on common/lib/xmodule/xmodule/modulestore/draft_and_published.py:
```diff
@@ -134,37 +138,58 @@ def _flag_publish_event(self, course_key):
+        log.info(
+            '%s parent updated to %s',
+            unicode(source_item.location),
+            unicode(old_parent_item.location)
```
`old_parent_item.location`=>`new_parent_location`

doctoryes commented on common/lib/xmodule/xmodule/modulestore/draft_and_published.py:
```diff
@@ -134,6 +138,59 @@ def _flag_publish_event(self, course_key):
+            new_parent_location (BlockUsageLocator)  : New parent block locator.
+            old_parent_location (BlockUsageLocator)  : Old parent block locator.
+            insert_at (int) : Insert item at the particular index in new parent.
+            user_id (int)   : User id
```
Please order these arguments in the same order as in the method parameters.

doctoryes commented on common/lib/xmodule/xmodule/modulestore/draft_and_published.py:
```diff
@@ -134,6 +138,59 @@ def _flag_publish_event(self, course_key):
+            user_id (int)   : User id
+
+        Returns:
+           location or None:    source item location if updated otherwise None.
```
`BlockUsageLocator or None:    Source item location if updated, otherwise None.`

doctoryes commented on common/lib/xmodule/xmodule/modulestore/draft_and_published.py:
```diff
@@ -134,6 +138,59 @@ def _flag_publish_event(self, course_key):
+            old_parent_item = self.get_item(old_parent_location)
+            new_parent_item = self.get_item(new_parent_location)
+        except ItemNotFoundError:
+            log.error('Unable to find the item %s', unicode(item_location))
```
The exception here could be triggered by item_location, old_parent_location, or new_parent_location. The logged error has a 67% chance of being misleading.

mushtaqak replied:
oh my bad, I will correct that

doctoryes commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split_draft.py:
```diff
@@ -444,8 +444,10 @@ def copy_from_published(root_block_id):
                 block = self._get_block_from_structure(new_structure, root_block_id)
                 for child_block_id in block.fields.get('children', []):
+                    item_location = location.course_key.make_usage_key(child_block_id.type, child_block_id.id)
+                    original_parent_location = location.course_key.make_usage_key(root_block_id.type, root_block_id.id)
```
Nit: This location def could be moved outside the `for` loop here.

doctoryes commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split_draft.py:
```diff
@@ -454,6 +456,27 @@ def copy_from_published(root_block_id):
+            user_id (int)   : User id
+        """
+        parent_block_keys = self._get_parents_from_structure(BlockKey.from_usage_key(item_location), course_structure)
+        for block_key in parent_block_keys:
```
If the code looks for the parent of `item_location`, in what case would it return more than one parent? Is this code to guard against items having more than one parent (directed acyclic graph)?

mushtaqak replied:
In split courses, I have seen a component in more than one parant verticals (so is the the move-discard-scenario) - where we get a component placed under more than one vertical's children list.

doctoryes replied:
Unfortunate! But believable...

mushtaqak commented:
jenkins run js

muzaffaryousaf commented:
@doctoryes Would you be able to continue review today ?

muzaffaryousaf commented on common/lib/xmodule/xmodule/modulestore/mongo/draft.py:
```diff
@@ -809,6 +816,27 @@ def delete_draft_only(root_location):
 
         delete_draft_only(location)
 
+    def update_parent_if_moved(self, original_parent_location, published_version, delete_draft_only, user_id):
```
@mushtaqak can we write unit test independently for this method ?

muhammad-ammar commented on common/lib/xmodule/xmodule/modulestore/draft_and_published.py:
```diff
@@ -134,6 +138,62 @@ def _flag_publish_event(self, course_key):
+            new_parent_item = self.get_item(new_parent_location)    # pylint: disable=no-member
+        except ItemNotFoundError as exception:
+            log.error('Unable to find the item : %s', exception.message)
+            return None
```
nit: don't need to return `None` explicitly, `None` is returned by default

doctoryes commented:
I agree with @muzaffaryousaf about documenting the move operation scenarios - but just ensure that everything documented has an accompanying test in both modulestores (Old Mongo/Split).

muhammad-ammar commented on common/lib/xmodule/xmodule/modulestore/mongo/draft.py:
```diff
@@ -809,6 +816,27 @@ def delete_draft_only(root_location):
+
+        Arguments:
+            original_parent_location (BlockUsageLocator)  : Original parent block locator.
+            published_version (dict)   : Published version of the course.
```
this will be a specific xblock published version?

muhammad-ammar commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split_draft.py:
```diff
@@ -454,6 +457,23 @@ def copy_from_published(root_block_id):
+        Arguments:
+            item_location (BlockUsageLocator)    : Locator of item.
+            original_parent_location (BlockUsageLocator)  : Original parent block locator.
+            course_structure (Structure)  : course structure of the course.
```
doc string needs to be updated. this is the structure of a particular xblock.

mushtaqak commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -761,16 +761,16 @@ def _move_item(source_usage_key, target_parent_usage_key, user, target_index=Non
-        # Add to new parent at particular location.
-        target_parent.children.insert(insert_at, source_item.location)
-        store.update_item(target_parent, user.id)
+        store.remove_update_item_parent(
```
ok I will do that as well

mushtaqak commented:
@doctoryes @muzaffaryousaf Please review.

Here is the [speadsheet](https://docs.google.com/spreadsheets/d/1MoEAUEuEl_Sq5pzDCv8MSKOyN9HhPLdW1ysrYqTumE0/edit#gid=0) that contains move scenarions

muzaffaryousaf commented on common/lib/xmodule/xmodule/modulestore/tests/test_mixed_modulestore.py:
```diff
@@ -1143,6 +1143,378 @@ def verify_get_parent_locations_results(self, expected_results):
+    def test_move_1_moved_1_unchanged(self, default_ms):
+        """
+        Test that when we move an item inside an old parent having multiple components, moved item is placed under new
+        parent while other component is still present in old parent.
```
need to reword this to make it understandable :)

muzaffaryousaf commented on common/lib/xmodule/xmodule/modulestore/tests/test_mixed_modulestore.py:
```diff
@@ -1143,6 +1143,378 @@ def verify_get_parent_locations_results(self, expected_results):
+        self.assertEqual(orig_display_name, reverted_problem2.display_name)
+
+    @ddt.data(ModuleStoreEnum.Type.mongo, ModuleStoreEnum.Type.split)
+    def a_test_move_1_moved_1_deleted(self, default_ms):
```
Mistake ?

muzaffaryousaf commented:
@mushtaqak can you get a green build ?

muzaffaryousaf commented:
@doctoryes do you want to take a look to the tests ?

mushtaqak commented on common/lib/xmodule/xmodule/modulestore/tests/test_mixed_modulestore.py:
```diff
@@ -1143,6 +1143,378 @@ def verify_get_parent_locations_results(self, expected_results):
+        self.assertNotIn(item_location, old_parent.children)
+
+    @ddt.data(ModuleStoreEnum.Type.mongo, ModuleStoreEnum.Type.split)
+    def test_update_item_parent(self, default_ms):
```
`default_ms ` is used at most places in this file as well others. But I can use `store_type` which is now used mostly to know store type.

Qubad786 commented on common/lib/xmodule/xmodule/modulestore/tests/test_mixed_modulestore.py:
```diff
@@ -1143,6 +1143,380 @@ def verify_get_parent_locations_results(self, expected_results):
+        # Now discard changes in old_parent_location i.e original parent.
+        self.store.revert_to_published(old_parent_location, self.user_id)
+
+        # Check that problem_item2 is also back in in vertical_x1a
```
in case someone missed the first one? 😄

Qubad786 commented on common/lib/xmodule/xmodule/modulestore/tests/test_mixed_modulestore.py:
```diff
@@ -1143,6 +1143,380 @@ def verify_get_parent_locations_results(self, expected_results):
+        new_parent_location = self.vertical_y1a
+        old_parent_location = self.vertical_x1a
+
+        # Move problem_x1a_1 from x1 to y1.
```
unnecessary comment, `L1405` is already serving the purpose.

mushtaqak commented:
@muzaffaryousaf @muhammad-ammar @Qubad786 All feedback addressed.

muhammad-ammar commented:
@mushtaqak lgtm 👍

muzaffaryousaf commented:
👍  LGTM
